Fix flaky Nav block user permissions e2e test

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description

The user permissions test for the Nav block have been flaky recently. I wondered why this was because when I committed they weren't - I ran several test runs on CI to confirm that fact.

After some digging I noticed that https://github.com/WordPress/gutenberg/commit/a92586c6ba83e735e102867bc69584b6af856643 moved the creation of the contributor user to be local to the `describe` block where the permissions tests are performed. However, when @ockham and I originally debugged this test (when it was created) we ascertained that the contributor user creation had to happen at the _top/root_ level `describe` block otherwise the test was unstable. We don't know _why_ that is but it is probably a quirk of Puppetter.

This PR moves the user creation back to the root level `describe` in the hope that will make the test stable again.

Closes https://github.com/WordPress/gutenberg/issues/37604

## How has this been tested?

Run the CI e2e tests and make sure that the x2 tests below do not fail:

* `shows a warning if user does not have permission to edit or update navigation menus`
* `shows a warning if user does not have permission to create navigation menus`

We should run the CI on this PR 5 times to be reasonably sure this fix does actually produce a more stable test. Please check off below when you see a valid test run:

- [x] [Test Run 1](https://github.com/WordPress/gutenberg/actions/runs/1707681638/attempts/1)
- [x] [Test Run 2](https://github.com/WordPress/gutenberg/actions/runs/1707681638/attempts/2) - unrelated failure.
- [x] [Test Run 3](https://github.com/WordPress/gutenberg/actions/runs/1707681638/attempts/3)
- [ ] Test Run 4
- [ ] Test Run 5

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [ ] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

getdave commented:
@talldan Could you do me a favour and (assuming you agree with the premise of this PR) restart the tests for runs 4 & 5, checking off s required in the PR desc.

If all good then I think we are ok to merge this one.

kevin940726 commented on packages/e2e-tests/specs/editor/blocks/navigation.test.js:
```diff
@@ -233,6 +245,10 @@ describe( 'Navigation', () => {
+
+		// As per the creation in the beforeAll() above, this
+		// action must be done at the root level describe() block.
+		await deleteUser( contributorUsername );
```
Shouldn't this best be put in `afterAll`?

getdave replied:
I'm pretty sure it is already in the `afterAll`...

<img width="876" alt="Screen Shot 2022-01-20 at 13 14 46" src="https://user-images.githubusercontent.com/444434/150345729-84396787-da40-44f9-89e6-02031cbb7282.png">

kevin940726 replied:
Oops, right, GitHub failed me 😂
![Screen Shot 2022-01-20 at 10 22 16 PM](https://user-images.githubusercontent.com/7753001/150356660-96c36cd3-729f-4eb0-bc11-7746a8b8553b.png)

talldan replied:
Wow, yep, I also completely missed that. 🤦

kevin940726 commented on packages/e2e-tests/specs/editor/blocks/navigation.test.js:
```diff
@@ -211,6 +211,18 @@ async function getNavigationMenuRawContent() {
+	let contributorPassword;
+
+	beforeAll( async () => {
+		// Creation of the contributor user **MUST** be at the top level describe block
+		// otherwise this test will become unstable. This action only happens once
+		// so there is no huge performance hit.
+		contributorPassword = await createUser( contributorUsername, {
+			role: 'contributor',
+		} );
```
Maybe there's a bug in the e2e helpers so that switching to a newly created user is unreliable? Otherwise, I can't think of any other reason why this has to be at the root level.

`loginUser` doesn't check if the submission is successful, it only tries to fill out the login form and press the login button. Perhaps we can add an assertion in the helper and/or throw a meaningful error message as to why it fails.

The same goes `createUser`. It doesn't check if the action is successful either. The password it returns is auto-generated before the submission, so checking its existence isn't enough either.

I'm guessing maybe one of these helpers failed at some point and we didn't catch it. The reason it's less unreliable in top-level describe block is still unknown though.

talldan replied:
It'd definitely be good to try to understand the root cause, as otherwise it'll be difficult to use this utility in other tests.

kevin940726 replied:
Probably a good opportunity to rewrite these utils with `rest` or something similar too 🙈 .

talldan replied:
Oh, I didn't realise it wasn't already using REST. Yeah, I definitely think that'd be a good idea.

getdave replied:
Sounds good. Are we able to ship this fix now though so we have test coverage and then circle back to making it REST?

kevin940726 replied:
Yes, absolutely. I have a draft PR opened at #38060 which might be helpful for that matter and I'll work on it next.

talldan commented on packages/e2e-tests/specs/editor/blocks/navigation.test.js:
```diff
@@ -211,6 +211,18 @@ async function getNavigationMenuRawContent() {
 // Disable reason - these tests are to be re-written.
 // eslint-disable-next-line jest/no-disabled-tests
 describe( 'Navigation', () => {
+	const contributorUsername = 'contributoruser';
```
I wonder if we should make the username unique by adding some randomness, as one of the reasons WordPress would prevent creation of a user is a duplicate username, and that might happen if clean up of this test fails, or possibly another test using the same username.

getdave replied:
I added randomness using lodash. Does that work?

kevin940726 replied:
This might be slightly over-engineered. I'd imagine a better solution would be having a proper error message if that happens. We are already running our tests in-band, so such things shouldn't happen. If they happen, it might be because our test cases aren't being correctly cleaned up, and we should fix them. But it's not that it's an anti-pattern or anything like that, I don't mind landing this 😅 .

talldan replied:
True, it shouldn't happen, but if it did I feel like it would still be really obscure and hard to track down. Better error messages would be good though.

I don't think `_.uniqueId` would do the job anyway, as that's just an incrementing integer. Another test in a new process would generate the same integer.

talldan commented:
I notice this is failing quite a lot, so I've skipped the test for now until this is ready to merge - https://github.com/WordPress/gutenberg/pull/38099

I'm happy to go ahead with this change as a temporary step to stabilise the test, but that[ first comment from @kevin940726](https://github.com/WordPress/gutenberg/pull/38025#discussion_r786377865) about the `deleteUser` call being in the wrong place needs to be fixed before it can be approved.

getdave committed:
Make unique contributor username

Acts a safeguard against duplicate users.

getdave committed:
Move user creation back to the top level

getdave commented:
> I notice this is failing quite a lot, so I've skipped the test for now until this is ready to merge - #38099

Nice one. Thanks.


> I'm happy to go ahead with this change as a temporary step to stabilise the test, but that[ first comment from @kevin940726](https://github.com/WordPress/gutenberg/pull/38025#discussion_r786377865) about the `deleteUser` call being in the wrong place needs to be fixed before it can be approved.

I _think_ @kevin940726 might be mistaken. It's already in the `afterAll`.

Also I rebased the branch and also added randomness to the username.

Let me know if this can merge or if there's more that needs to happen.

getdave commented:
Just re-running this a few times to ensure it's still stable. @talldan when you come online if this is still ✅ feel free to merge.
